README.md
# Rollup plugin to serve the bundle

<a href="LICENSE">
  <img src="https://img.shields.io/badge/license-MIT-brightgreen.svg" alt="Software License" />
</a>
<a href="https://github.com/thgh/rollup-plugin-serve/issues">
  <img src="https://img.shields.io/github/issues/denyncrawford/deno-rollup-plugin-serve.svg" alt="Issues" />
</a>
<a href="https://github.com/standard/ts-standard/">
  <img src="https://img.shields.io/badge/code%20style-standard-brightgreen.svg" alt="Typescript Style Guide" />
</a>
<a href="https://deno.land/x/deno-rollup-plugin-serve">
  <img src="https://img.shields.io/badge/deno-^1.8.1-informational.svg?style=flat-squar" alt="Deno.land" />
</a>
<a href="https://github.com/denyncrawford/deno-rollup-plugin-serve/releases">
  <img src="https://img.shields.io/github/release/denyncrawford/deno-rollup-plugin-serve.svg" alt="Latest Version" />
</a>

[![nest badge](https://nest.land/badge.svg)](https://nest.land/package/your-module)

## Overview

This modules made for [deno-rollup](https://github.com/cmorten/deno-rollup) ***Next-generation ES module bundler for Deno ported from Rollup.***

**deno-rollup-plugin-serve** is a Deno rebuilt/port from [rollup-plugin-serve](https://www.npmjs.com/package/rollup-plugin-serve). It creates a development server for testing your front-end apps.

## Import
```typescript
# Deno Rollup v2.36.1+0.1.2+
import serve from 'https://deno.land/x/drollup_plugin_serve@1.1.0+0.1.3/mod.ts'
// or
import serve from 'http://x.nest.land/deno-rollup-plugin-serve@4.1.0/mod.ts'
```

## Usage
```js
// rollup.config.js
import serve from '...'

export default {
  input: 'src/main.js',
  output: {
    file: 'dist/bundle.js',
    format: ...
  },
  plugins: [
    serve('dist')
  ]
}
```

### Options

By default it serves the current project folder. Change it by passing a string:
```js
serve('public')    // will be used as contentBase

// Default options
serve({
  // Launch in browser (default: false)
  open: true,

  // Page to navigate to when opening the browser.
  // Will not do anything if open=false.
  // Remember to start with a slash.
  openPage: '/different/page',

  // Show server address in console (default: true)
  verbose: false,

  // Folder to serve files from
  contentBase: '',

  // Multiple folders to serve from
  contentBase: ['dist', 'static'],

  // Set to true to return index.html (200) instead of error page (404)
  historyApiFallback: false,

  // Path to fallback page
  historyApiFallback: '/200.html',

  // Options used in setting up server
  host: 'localhost',
  port: 10001,

  // By default server will be served over HTTP (https: false). It can optionally be served over HTTPS
  https: {
    key: fs.readFileSync('/path/to/server.key'),
    cert: fs.readFileSync('/path/to/server.crt'),
    ca: fs.readFileSync('/path/to/ca.pem')
  },

  //set headers
  headers: {
    'Access-Control-Allow-Origin': '*',
    foo: 'bar'
  },

  // set custom mime types, usage https://github.com/broofa/mime#mimedefinetypemap-force--false
  mimeTypes: {
    'application/javascript': ['js_commonjs-proxy']
  }

  // execute function after server has begun listening
  onListening: function (server) {
    const address = server.address()
    const host = address.address === '::' ? 'localhost' : address.address
    // by using a bound function, we can access options as `this`
    const protocol = this.https ? 'https' : 'http'
    console.log(`Server listening at ${protocol}://${host}:${address.port}/`)
  }
})
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Contributions and feedback are very welcome.

This project aims to stay lean and close to standards. New features should encourage to stick to standards. Options that match the behaviour of [webpack-dev-server](https://webpack.js.org/configuration/dev-server/#devserver) are always ok.

## Credits

- [Thomas Ghysels](https://github.com/thgh)
- [Miguel Rangel][link-author-2]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.

[link-author]: https://github.com/thgh
[link-author-2]: https://github.com/denyncrawford
[link-contributors]: ../../contributors
[rollup-plugin-serve]: https://www.npmjs.com/package/rollup-plugin-serve


CHANGELOG.md
# Changelog

All notable changes to `deno-rollup-plugin-serve` including the previos at the node version will be documented in this file.

## [1.1.0+0.1.3]

### Fixed
- Sending binary data as files correctly.
- Making all options optional
- Fixed basepath not found issue

### Added

- Types and Interfaces

## [1.1.0+0.1.2]
### Added
- Routes without ***/*** checks if they're folders to resolve index.html

### Fixed
- Deno error handling.
- historyApiFallback now works and can handle boolean or strings

## [1.1.0+0.1.0]
### Ported
- Ported all functionalities from Node to Deno.

## [1.1.0] - 2020-11-01
### Added
- Add `onListening` hook #69 @filoxo
- Show error message when port is in use #60 @jaeh

## [1.0.4] - 2020-08-28
### Added
- Add `mimeTypes` option #58 @GerardRodes

### Fixed
- Allow opening the browser even when verbose mode is disabled #64 @Richienb

## [1.0.3] - 2020-07-21
### Fixed
- Fix path.normalize error on Windows

## [1.0.2] - 2020-07-17
### Fixed
- Fix path traversal issue

## [1.0.1] - 2019-01-27
### Added
- Add Intellisense support #34

### Fixed
- Set minimum version for `mime` package

## [1.0.0] - 2019-01-11
### Fixed
- Update `ongenerate` to `generateBundle`

### Removed
- Remove built-in `favicon.ico` #20

## [0.6.1] - 2018-12-23
### Added
- Add support for `rollup --watch` (Release http server on rollup reload)

## [0.5.0] - 2018-09-18
### Added
- Allow to override path for historyApiFallback option
- Option `openPage`

### Fixed
- Fix host option

## [0.4.2] - 2017-08-25
### Added
- Option `https`

## [0.4.1] - 2017-08-16
### Added
- Option `headers`

### Fixed
- Close the server when a termination signal is encountered

## [0.4.0] - 2017-07-02
### Fixed
- Various fixes to contentBase handling

## [0.3.0] - 2017-04-07
### Changed
- Allow to pass array as contentBase

## [0.2.0] - 2017-04-07
### Changed
- Show console message only once

## [0.1.0] - 2016-09-29
### Added
- Option `open` open the project in browser
- Option `historyApiFallback`
- Default green favicon

## Changed
- Option `root` is now `contentBase`

## [0.0.1] - 2016-09-24
### Added
- Initial version

[Unreleased]: https://github.com/thgh/rollup-plugin-serve/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.6.1...v1.0.0
[0.6.1]: https://github.com/thgh/rollup-plugin-serve/compare/v0.5.0...v0.6.1
[0.5.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.4.2...v0.5.0
[0.4.2]: https://github.com/thgh/rollup-plugin-serve/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/thgh/rollup-plugin-serve/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.0.1...v0.1.0
[0.0.1]: https://github.com/thgh/rollup-plugin-serve/releases


test/rollup.config.js
import serve from '../mod.ts'

export default { 
  input: './entry.js',
  plugins: [
    serve({
      open: true,
      port: 3000,
      openPage: '/frames.html',
      historyApiFallback: true,
      contentBase: ['.', 'base1', 'base2'],
      onListening({protocol, host, port}) {
        // by using a bound function, we can access options as `this`
        console.log(`Server listening at ${protocol}://${host}:${port}/`)
      }
    })
  ],
  output: {
    file: 'dest.js',
    format: 'iife'
  },
  watch: {
    clearScreen: true,
  }
}


test/rollup.config.ts
import serve from '../mod.ts'

export default { 
  input: './entry.js',
  plugins: [
    serve({
      open: true,
      port: 3000,
      openPage: '/frames.html',
      historyApiFallback: true,
      contentBase: ['.', 'base1', 'base2'],
      onListening({protocol, host, port}: { protocol: string; host: string, port: string }) {
        // by using a bound function, we can access options as `this`
        console.log(`Server listening at ${protocol}://${host}:${port}/`)
      }
    })
  ],
  output: {
    file: 'dest.js',
    format: 'iife'
  },
  watch: {
    clearScreen: true,
  }
}


src/types.ts
import { MimeTypeMap } from 'https://deno.land/x/mimetypes@v1.0.0/src/mime.ts';
import { HTTPSOptions } from 'https://deno.land/std@0.92.0/http/server.ts';

export interface ServeOptions<T = unknown> {
  contentBase: Array<string>,
  port: number,
  host: string,
  headers: Record<string, string>,
  https?: HTTPSOptions,
  openPage: string,
  onListening: (adress: Record<string, string | number>) => void,
  mimeTypes?: MimeTypeMap,
  defaultType: string,
  verbose: boolean,
  open: boolean,
  historyApiFallback?: (string | boolean)
}

export type Defined<T> = Exclude<T, undefined>;

export type Inner<T extends ServeOptions<unknown>> = T extends ServeOptions<infer X> ? X
  : never;


export type ReadReturn = {
  err: ErrorConstructor | null,
  filePath: string,
  size: string | null,
  content: Uint8Array | null
}

src/index.ts
const { readFile } = Deno
import { ServeOptions, ReadReturn, Inner } from './types.ts'
import { mime } from 'https://deno.land/x/mimetypes@v1.0.0/src/mime.ts';
import { open } from "https://deno.land/x/opener/mod.ts";
import { 
  serve, 
  serveTLS,
  ServerRequest, 
  Response, 
  Server } from "https://deno.land/std@0.92.0/http/server.ts";
import { normalize, resolve } from 'https://deno.land/std@0.92.0/path/mod.ts'
import type {
  Plugin,
} from 'https://deno.land/x/drollup@2.42.3+0.17.1/mod.ts'

/**
 * Serve your rolled up bundle like webpack-dev-server
 * @param {ServeOptions|string|string[]} options
 */

const decoder = new TextDecoder()

class BuildServer <Options extends Record<string, ServeOptions<unknown>>> {
  options: ServeOptions = {
    contentBase: [''],
    port: 10001,
    host: 'localhost',
    headers: {},
    openPage: '',
    open: true,
    defaultType: 'text/plain',
    verbose: true,
    onListening () {}
  };

  server: Server;
  #p?: Promise<void>;
  first = true
  constructor (initOptions: string | Array<string> | { [K in keyof Options]?: Inner<Options[K]> } = [''] ) {
    if (Array.isArray(initOptions) || typeof initOptions === 'string') {
      this.options.contentBase = typeof initOptions === 'string' ? [initOptions] : initOptions;
    }
    Object.assign(this.options, initOptions)
    this.options.contentBase = typeof this.options.contentBase === 'string' ? [this.options.contentBase] : this.options.contentBase;
    if (this.options?.mimeTypes) {
      mime.define(this.options?.mimeTypes, true);
    }
    this.server = !this.options.https
    ? serve({port: this.options.port || 10001, hostname: this.options.host})
    : serveTLS(this.options.https)
  }

  async requestHandler(req: ServerRequest) {
    const response: Response = {
      headers: new Headers(this.options.headers)
    }; 
    // Remove querystring
    const unsafePath = decodeURI(req.url.split('?')[0]);
    // Don't allow path traversal
    const urlPath = normalize(unsafePath)
    const { content, err, filePath, size } = await readFileFromContentBase(this.options.contentBase, urlPath)
    if (!err && content) return req.respond(this.found(response, filePath, content, size ? size : ''))
    if (err && err.name !== 'NotFound') {
      response.status = 500;
      response.body = '500 Internal Server Error' +
        '\n\n' + filePath +
        '\n\n' + /*Object.values(err).join('\n')*/ err +
        '\n\n(rollup-plugin-serve)';
      return req.respond(response)
    }
    if (this.options.historyApiFallback) {         
      const fallbackPath = typeof this.options.historyApiFallback === 'string' ? this.options.historyApiFallback : '/index.html'
      const { content: bContent, err: bError, filePath: bFilepath, size: bSize } = await readFileFromContentBase(this.options.contentBase, fallbackPath)
        if (bError) {
          return req.respond(this.notFound(response, filePath))
        } else {
          return req.respond(this.found(response, bFilepath, bContent || new Uint8Array(0), bSize ? bSize : ''))
        }
    } else {
      return req.respond(this.notFound(response, filePath))
    }
  }
  async serve() {
    if (this.options.onListening) this.options.onListening({
      port: this.options.port, 
      host: this.options.host,
      protocol: this.options.https ? 'https' : 'http'
    })
    for await(const req of this.server) {
      this.requestHandler(req);
    }
  }

  notFound(response: Response, filePath: string): Response {
    response.status = 404;
    response.body = '404 Not Found' +
      '\n\n' + filePath +
      '\n\n(rollup-plugin-serve)';
    return response;
  }
  
  found(response: Response, filePath: string, content: Uint8Array, size: string): Response {
    response.status = 200;
    if (response.headers) {
      response.headers.append("content-length", size);
      response.headers.append('Content-Type', mime.getType(filePath) || this.options.defaultType)
    }
    response.body = content;
    return response;
  }
  
  green(text: string) {
    return '\u001b[1m\u001b[32m' + text + '\u001b[39m\u001b[22m'
  }

  rollup() {
    //if (this.server) this.closeServerOnTermination()
    const url = (this.options.https ? 'https' : 'http') + '://' + (this.options.host || 'localhost') + ':' + this.options.port
    let first = true;
    const options = this.options,
    green = this.green;
    this.serve()
    return {
      name: 'serve',
      generateBundle() {
        if (first) {
          first = false
          // Log which url to visit
          if (options.verbose !?? false) {
            options.contentBase.forEach((base: string) => {
              console.log(green(url) + ' -> ' + resolve(base));
            })
          }
          // Open browser
          if (options.open) {
            /* @ts-ignore */
            if (/https?:\/\/.+/.test(options.openPage)) {
              /* @ts-ignore */
              open(options.openPage)
            } else {
              open(url + options.openPage)
            }
          }
        }
      }
    }
  }
}

const readFileFromContentBase = async (contentBase:Array<string>, urlPath: string): Promise<ReadReturn> => {
  let filePath = resolve(contentBase[0] || '.', '.' + urlPath)
  // Load index.html in directories

  if (urlPath.endsWith('/')) {
    filePath = resolve(filePath, 'index.html')
  } 
  
  if (!urlPath.endsWith('/') && !urlPath.includes('.')) {
    try {
      const fileInfo = await Deno.stat(filePath)
      filePath = fileInfo.isFile ? filePath : resolve(filePath, 'index.html')
    } catch (err) {
      filePath
    }
  }

  // Try Read
  try {
    const [content, fileInfo] = await Promise.all([
      readFile(filePath),
      Deno.stat(filePath),
    ]);
    return {
      err: null,
      filePath,
      size: fileInfo.size.toString(),
      content      
    }
  } catch (err) {
    if (err && contentBase.length > 1) return readFileFromContentBase(contentBase.slice(1), urlPath)
    // We know enough
    else return {
      err,
      filePath,
      size: null,
      content: null
    }
  }
}

type Options = Record<string, ServeOptions<unknown>>

export default (initOptions: string | Array<string> | { [K in keyof Options]?: Inner<Options[K]> } = ['']): Plugin => {
  const server = new BuildServer(initOptions);
  const plugin = server.rollup();
  return plugin;
}

/**
 * @typedef {Object} ServeOptions
 * @property {boolean} [open=false] Launch in browser (default: `false`)
 * @property {string} [openPage=''] Page to navigate to when opening the browser. Will not do anything if `open` is `false`. Remember to start with a slash e.g. `'/different/page'`
 * @property {boolean} [verbose=true] Show server address in console (default: `true`)
 * @property {string|string[]} [contentBase=''] Folder(s) to serve files from
 * @property {string|boolean} [historyApiFallback] Path to fallback page. Set to `true` to return index.html (200) instead of error page (404)
 * @property {string} [host='localhost'] Server host (default: `'localhost'`)
 * @property {number} [port=10001] Server port (default: `10001`)
 * @property {function} [onListening] Execute a function when server starts listening for connections on a port
 * @property {ServeOptionsHttps} [https=false] By default server will be served over HTTP (https: `false`). It can optionally be served over HTTPS
 * @property {{[header:string]: string}} [headers] Set headers
 */

/**
 * @typedef {Object} ServeOptionsHttps
 * @property {string|Buffer|Buffer[]|Object[]} key
 * @property {string|Buffer|Array<string|Buffer>} cert
 * @property {string|Buffer|Array<string|Buffer>} ca
 * @see https.ServerOptions
 */